### PR TITLE
Block cache analyzer: Compute correlation of features and human readable trace file.

### PR DESCRIPTION
**Pull request by HaoyuHuang**
- Compute correlation between a few features and predictions, e.g., number of accesses since the last access vs number of accesses till the next access on a block. 
- Output human readable trace file so python can consume it. 

Test plan: make clean && USE_CLANG=1 make check -j32

### CODE REVIEW

File: tools/block_cache_trace_analyzer.cc [hunk 1]
```diff
@@ -279,6 +304,34 @@ double percent(uint64_t numerator, uint64_t denomenator) {
...
+  return adjusted_time_stats;
+}
+
+int get_free_memory() {
+  FILE* fpipe = nullptr;
+  std::string command = "free -g | grep \"Mem\" | grep -o '[^ ]*$'";
+  char c = 0;
+  std::string output;
+  if (0 == (fpipe = (FILE*)popen(command.c_str(), "r"))) {
+    fprintf(stderr, "popen() failed.");
```
Comment (riversand963) [inline tools/block_cache_trace_analyzer.cc]: Print a newline.

File: tools/block_cache_trace_analyzer.cc [hunk 2]
```diff
@@ -314,13 +372,293 @@ void BlockCacheTraceAnalyzer::WriteMissRatioCurves() const {
...
+      (*label_predictions)[label].num_accesses_till_next_access.push_back(
+          num_accesses_since_last_access);
+      (*label_predictions)[label].elapsed_time_till_next_access.push_back(
+          elapsed_time_since_last_access);
+    }
+  }
+}
+
+void BlockCacheTraceAnalyzer::WriteMissRatioTimeline(uint64_t time_unit) const {
+  if (!cache_simulator_) {
```
Comment (riversand963) [inline tools/block_cache_trace_analyzer.cc]: If you are not going to print some extra logging, I think you can combine these two early-return cases.

File: tools/block_cache_trace_analyzer.cc [hunk 3]
```diff
@@ -7,11 +7,16 @@
 #ifdef GFLAGS
 #include "tools/block_cache_trace_analyzer.h"
 
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply tools/block_cache_trace_analyzer.cc]: Thanks. I changed to use cstdio, cstdlib.

File: tools/block_cache_trace_analyzer.cc [hunk 4]
```diff
@@ -279,6 +304,34 @@ double percent(uint64_t numerator, uint64_t denomenator) {
...
+  }
+  return adjusted_time_stats;
+}
+
+int get_free_memory() {
+  FILE* fpipe = nullptr;
+  std::string command = "free -g | grep \"Mem\" | grep -o '[^ ]*$'";
+  char c = 0;
+  std::string output;
+  if (0 == (fpipe = (FILE*)popen(command.c_str(), "r"))) {
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply tools/block_cache_trace_analyzer.cc]: As discussed offline, I removed this function and will add a GetFreeMemory function in env.

File: tools/block_cache_trace_analyzer.cc [hunk 5]
```diff
@@ -279,6 +304,34 @@ double percent(uint64_t numerator, uint64_t denomenator) {
...
+  for (auto const& time : time_stats) {
+    adjusted_time_stats[static_cast<uint64_t>(time.first / time_unit)] +=
+        time.second;
+  }
+  return adjusted_time_stats;
+}
+
+int get_free_memory() {
+  FILE* fpipe = nullptr;
+  std::string command = "free -g | grep \"Mem\" | grep -o '[^ ]*$'";
```
[the remark this answers is not in the document]
Comment (HaoyuHuang) [reply tools/block_cache_trace_analyzer.cc]: As discussed offline, I removed this function and will add a GetFreeMemory function in env. I will use sysinfo for Linux platforms.

File: tools/block_cache_trace_analyzer.cc (hunk 1 above)
Comment (HaoyuHuang) [reply tools/block_cache_trace_analyzer.cc]: Thanks. I removed this function.

File: tools/block_cache_trace_analyzer.cc (hunk 2 above)
Comment (HaoyuHuang) [reply tools/block_cache_trace_analyzer.cc]: Sure. I combined these two.

**Pull request closed without merging**